root/static/scripts/account/components/EditProfileForm.js

```diff
@@ -0,0 +1,334 @@
+  options: [
+    {label: N_l('Male'), value: 1},
+    {label: N_l('Female'), value: 2},
+    {label: N_l('Other'), value: 3},
+  ],
+};
+
+const fluencyOptions = {
+  grouped: false,
+  options: [
+    {label: N_l('Basic'), value: 'basic'},
+    {label: N_l('Intermediate'), value: 'intermediate'},
+    {label: N_l('Advanced'), value: 'advanced'},
+    {label: N_l('Native'), value: 'native'},
+  ],
+};
+
+const genderFieldLens: Lens<EditProfileFormT, number> =
+  compose3(prop('field'), prop('gender_id'), prop('value'));
+
+class EditProfileForm extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+    this.state = {form: props.form, languageOptions: props.language_options};
+    this.handleGenderChange = this.handleGenderChange.bind(this);
+    this.handleLanguageChange = this.handleLanguageChange.bind(this);
+    this.handleFluencyChange = this.handleFluencyChange.bind(this);
+    this.removeLanguage = this.removeLanguage.bind(this);
+    this.addLanguage = this.addLanguage.bind(this);
+  }
+
+  componentDidMount() {
+    const $ = require('jquery');
+    const MB = require('../../common/MB');
+    require('../../common/MB/Control/Autocomplete');
+    // $FlowFixMe - EntityAutocomplete is missing in MB.Control
+    MB.Control.EntityAutocomplete({inputs: $('span.area.autocomplete')});
+  }
+
+  handleGenderChange: (e: SyntheticEvent<HTMLSelectElement>) => void;
+
+  handleGenderChange(e: SyntheticEvent<HTMLSelectElement>) {
+    const selectedGender = e.currentTarget.value;
+    this.setState(prevState => ({
+      form: set(
+        genderFieldLens,
+        Number.parseInt(selectedGender, 10),
+        prevState.form,
+      ),
+    }));
+  }
+
+  handleLanguageChange: (e: SyntheticEvent<HTMLSelectElement>) => void;
+
+  handleLanguageChange(e: SyntheticEvent<HTMLSelectElement>) {
+    const languageIndex = Number.parseInt(e.currentTarget.name.split('.')[2], 10);
+    const selectedLanguage = Number.parseInt(e.currentTarget.value, 10);
+    this.setState(function (prevState) {
+      const newState = {form: prevState.form};
+      newState.form.field.languages.field[languageIndex].field.language_id.value = selectedLanguage;
+      return newState;
+    });
+  }
+
+  handleFluencyChange: (e: SyntheticEvent<HTMLSelectElement>) => void;
+
+  handleFluencyChange(e: SyntheticEvent<HTMLSelectElement>) {
+    const languageIndex = Number.parseInt(e.currentTarget.name.split('.')[2], 10);
+    const selectedFluency = e.currentTarget.value;
+    this.setState(function (prevState) {
+      const newState = {form: prevState.form};
+      newState.form.field.languages.field[languageIndex].field.fluency.value = (selectedFluency: any);
+      return newState;
+    });
+  }
+
+  removeLanguage: (languageIndex: number) => void;
+
+  removeLanguage(languageIndex) {
+    this.setState(function (prevState) {
+      const newState = {form: prevState.form};
+      // $FlowFixMe - property splice is missing in $ReadOnlyArray
```

Comment:
Same here, https://github.com/metabrainz/musicbrainz-server/blob/60e0771/root/static/scripts/work.js#L113 would be a good example of removing a repeatable field without mutating the current state.

Reply:
Rewritten at https://github.com/metabrainz/musicbrainz-server/pull/730/commits/cdd603772f2625eace9cab4aed0d8349cf5a5918#diff-1fe48ebbd702a12c508208ed08f8e4f9R166